### PR TITLE
Semver

### PR DESCRIPTION
**Pull request by felipesere**
First step in the direction.
Heavily inspired by package-arg and semver_parser itself.

The parsing is probably not 100% spec compliant, as it needs to allow hyphens in certain places.

### CODE REVIEW

File: crates/oro-node-semver/src/version_req.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+where
+    E: ParseError<&'a str>,
+{
+    context(
+        "predicate alternatives",
+        alt((hypenated_with_only_major, full_version_range)),
+    )(input)
+}
+
+fn hypenated_with_only_major<'a, E>(input: &'a str) -> IResult<&'a str, Vec<Predicate>, E>
```
Comment (felipesere) [inline crates/oro-node-semver/src/version_req.rs:93]: I am a little worried I am going off track here is this looks if-else heavy.

Are there better strategies to parsing this?

File: crates/oro-node-semver/src/version_req.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+    }
+
+    range_parse_tests![         // input          // parsed and then `to_string`ed
+        parse_a_range =>        ["1.0.0 - 2.0.0", ">=1.0.0 <=2.0.0"],
+        only_major_versions =>  ["1 - 2",         ">=1.0.0 <3.0.0"],
+        only_major_and_minor => ["1.0 - 2.0",     ">=1.0.0 <2.1.0"],
+    ];
+    /*
+    ["1.0.0", "1.0.0", { loose: false }],
```
Comment (felipesere) [inline crates/oro-node-semver/src/version_req.rs:255]: I took the entire list from NPM :)

File: crates/oro-node-semver/src/version_req.rs [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+            )),
+            |(major, _, maybe_minor, _, _)| Range::Closed {
+                upper: upper_bound(major, maybe_minor),
+                lower: lower_bound(major, maybe_minor),
+            },
+        ),
+    )(input)
+}
+
+fn lower_bound(major: usize, maybe_minor: Option<usize>) -> Predicate {
```
Comment (zkat) [inline crates/oro-node-semver/src/version_req.rs:177]: Same as with Version, this should be u64, with an additional size check

File: crates/oro-node-semver/src/version_req.rs [hunk 4]
```diff
@@ -0,0 +1,636 @@
...
+    Predicate {
+        operation: Operation::GreaterThanEquals,
+        version: (major, maybe_minor.unwrap_or(0), 0).into(),
+    }
+}
+
+fn upper_bound(major: usize, maybe_minor: Option<usize>) -> Predicate {
```
Comment (zkat) [inline crates/oro-node-semver/src/version_req.rs:184]: ditto

File: crates/oro-node-semver/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    context(
+        "version core",
+        map(
+            tuple((number, tag("."), number, tag("."), number)),
+            |(major, _, minor, _, patch)| (major, minor, patch),
+        ),
+    )(input)
+}
+
+// I believe build, pre_release, and identifier are not 100% spec compliant.
```
[the remark this answers is not in the document]
Comment (felipesere) [reply crates/oro-node-semver/src/lib.rs:154]: I use the same parser & type for pre_release and identifier, which is not 100% correct.

**Pull request merged**